numba/targets/arrayobj.py

```diff
@@ -1702,6 +1702,31 @@ def imp(ary):
     return res
 
 
+@overload(np.clip)
+def np_clip(a, a_min, a_max, out=None):
+    def np_clip_impl(a, a_min, a_max, out=None):
+        if a_min is None and a_max is None:
+            raise ValueError("array_clip: must set either max or min")
+        if out is None:
+            out = np.empty_like(a)
+        for i in range(len(a)):
```

Comment:
Note to self: use `for index, val in np.ndenumerate(a):` and add multi-dimensional tests.

Reply:
Assuming that fixing `np.ndenumerate(scalar)` may well be complicated, could the implementation just branch to an function that specifically deals with scalars as a workaround? e.g.

```python
@overload(np.clip)
def np_clip(a, a_min, a_max, out=None):
    if isinstance(a, types.Number):
       def np_clip_impl(a, a_min, a_max, out=None):
            # scalar impl
    elif isinstance(a, types.Array):
        def np_clip_impl(a, a_min, a_max, out=None):
            # array impl
    return np_clip_impl
```